google-ads-examples/src/main/java/com/google/ads/googleads/examples/reporting/ParallelReportDownload.java

```diff
@@ -0,0 +1,249 @@
+    try (GoogleAdsServiceClient serviceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // IMPORTANT: You should avoid hitting the same customer ID in parallel. There are rate limit
+      // errors at the customer ID level, but not at the developer token level. Rate limits are
+      // retried by the client library with exponential back-off, and therefore would reduce overall
+      // throughput.
+      for (String gaqlQuery : GAQL_QUERY_STRINGS) {
+        // Uses a list of futures to make sure that we wait for this report to complete on all
+        // customer IDs before proceeding. The Future data type here is just for demonstration.
+        List<ListenableFuture<ReportSummary>> futures = new ArrayList<>();
+
+        // Calls the API for each customer ID on the current report.
+        for (Long customerId : customerIds) {
+          // Uses the gRPC asynchronous API to download the reports in parallel. This saves having
+          // to create/manage our own thread pool.
+          ResponseCountingObserver responseObserver = new ResponseCountingObserver(customerId);
+
+          // Starts the report download in a background thread.
+          serviceClient
+              .searchStreamCallable()
+              .call(
+                  SearchGoogleAdsStreamRequest.newBuilder()
+                      .setCustomerId(customerId.toString())
+                      .setQuery(gaqlQuery)
+                      .build(),
+                  responseObserver);
+
+          // Stores a future to retrieve the results.
+          futures.add(responseObserver.asFuture());
+        }
+
+        // Waits for all pending requests to the current set of customer IDs to complete.
+        //
+        // This is a naive implementation for illustrative purposes. It is possible to optimize the
+        // utilization of each customer ID by providing a queue of work (or similar). However, this
+        // would complicate the example code and so is omitted here.
+        List<ReportSummary> results =
+            Futures.allAsList(futures)
+                .get(DEADLINE_MILLIS - System.currentTimeMillis(), TimeUnit.MILLISECONDS);
+
+        System.out.println("Report results for query: " + gaqlQuery);
+        results.forEach(System.out::println);
+      }
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException(e);
+    } catch (ExecutionException | TimeoutException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** An observer which records a simple count of the result rows received. */
+  private static class ResponseCountingObserver
+      implements ResponseObserver<SearchGoogleAdsStreamResponse> {
+
+    private final long customerId;
+    private final SettableFuture<ReportSummary> future = SettableFuture.create();
+    private AtomicLong numRecords = new AtomicLong(0);
```

Comment:
`numResponses` would be less confusing. I misread this initially as the number of _rows_ across the responses.